stack-graphs/src/serde.rs

```diff
@@ -0,0 +1,588 @@
+            _source: &Handle<Node>,
+            _sink: &Handle<Node>,
+        ) -> bool {
+            true
+        }
+
+        fn include_partial_path(
+            &self,
+            _graph: &StackGraph,
+            _paths: &PartialPaths,
+            _path: &PartialPath,
+        ) -> bool {
+            true
+        }
+    }
+
+    // Filter implementation that includes everything.
+    pub struct NoFilter;
+
+    impl Filter for NoFilter {
+        fn include_file(&self, _graph: &StackGraph, _file: &Handle<File>) -> bool {
+            true
+        }
+
+        fn include_node(&self, _graph: &StackGraph, _node: &Handle<Node>) -> bool {
+            true
+        }
+
+        fn include_edge(
+            &self,
+            _graph: &StackGraph,
+            _source: &Handle<Node>,
+            _sink: &Handle<Node>,
+        ) -> bool {
+            true
+        }
+
+        fn include_partial_path(
+            &self,
+            _graph: &StackGraph,
+            _paths: &PartialPaths,
+            _path: &PartialPath,
+        ) -> bool {
+            true
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Default, Clone)]
+pub struct StackGraph {
+    pub files: Files,
+    pub nodes: Nodes,
+    pub edges: Edges,
+}
+
+#[derive(Debug, Error, PartialEq, Eq)]
+pub enum Error {
+    #[error("failed to load file `{0}`")]
+    FileNotFound(String),
+
+    #[error("duplicate file `{0}`")]
+    FileAlreadyPresent(String),
+
+    #[error("failed to locate node `{0}` in graph")]
+    NodeNotFound(u32),
+
+    #[error("no file data for node `{0}`")]
+    NoFileData(u32),
+
+    #[error("node `{0}` is an invalid node")]
+    InvalidNode(u32),
+}
+
+impl StackGraph {
+    pub fn from_graph<'a>(graph: &crate::graph::StackGraph) -> Self {
+        Self::from_graph_filter(graph, &NoFilter)
+    }
+
+    pub fn from_graph_filter<'a>(graph: &crate::graph::StackGraph, filter: &'a dyn Filter) -> Self {
+        let files = graph.filter_files(filter);
+        let nodes = graph.filter_nodes(filter);
+        let edges = graph.filter_edges(filter);
+        Self {
+            files,
+            nodes,
+            edges,
+        }
+    }
+
+    pub fn load_into(&self, graph: &mut crate::graph::StackGraph) -> Result<(), Error> {
+        self.load_files(graph)?;
+        self.load_nodes(graph)?;
+        self.load_edges(graph)?;
+        Ok(())
+    }
+
+    fn load_files(&self, graph: &mut crate::graph::StackGraph) -> Result<(), Error> {
+        // we check if any of the files we are about to introduce, are
+        // already present in this graph, to avoid accidental merges
+        if let Some(f) = self
+            .files
+            .data
+            .iter()
+            .find(|f| graph.get_file(f.as_str()).is_some())
+        {
+            return Err(Error::FileAlreadyPresent(f.to_owned()));
+        }
+
+        for f in self.files.data.iter() {
+            graph.add_file(f.as_str()).unwrap();
+        }
```

Comment:
the reasoning for this structuring is because `graph.add_file` is a mutating operation. if we exit midway through the iteration with `graph.add_file(..).map_err(..)?`, the graph might be left in a partially merged state. what do you think?

Reply:
You're correct that we'd leave the stack graph in an undefined state. At the moment none of the code really makes an effort to prevent this for mutable operations. Given that, I'd say cleaner code is preferred here.